core/src/services/mysql/backend.rs

```diff
@@ -0,0 +1,317 @@
+}
+
+impl Debug for Adapter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Adapter")
+            .field("connection_pool", &self.connection_pool)
+            .field("config", &self.config)
+            .field("table", &self.table)
+            .field("key_field", &self.key_field)
+            .field("value_field", &self.value_field)
+            .finish()
+    }
+}
+
+#[async_trait]
+impl kv::Adapter for Adapter {
+    fn metadata(&self) -> kv::Metadata {
+        kv::Metadata::new(
+            Scheme::Mysql,
+            &self.table,
+            Capability {
+                read: true,
+                write: true,
+                ..Default::default()
+            },
+        )
+    }
+
+    async fn get(&self, path: &str) -> Result<Option<Vec<u8>>> {
+        let query = format!(
+            "SELECT `{}` FROM `{}` WHERE `{}` = :path LIMIT 1",
+            self.value_field, self.table, self.key_field
+        );
+        let mut conn = self.connection_pool.get_conn().await.map_err(|err| {
+            Error::new(ErrorKind::Unexpected, "connection failed")
+                .with_context("service", Scheme::Mysql)
+                .set_source(err)
+        })?;
+        let statement = conn.prep(query).await.map_err(|err| {
+            Error::new(ErrorKind::Unexpected, "prepare statement failed")
+                .with_context("service", Scheme::Mysql)
+                .set_source(err)
+        })?;
+        let result: Option<Vec<u8>> = conn
+            .exec_first(
+                statement,
+                params! {
+                    "path" => path,
+                },
+            )
+            .await
+            .map_err(|err| {
+                Error::new(ErrorKind::Unexpected, "delete failed")
+                    .with_context("service", Scheme::Mysql)
+                    .set_source(err)
+            })?;
+        match result {
+            Some(v) => Ok(Some(v)),
+            None => Ok(None),
+        }
+    }
+
+    async fn set(&self, path: &str, value: &[u8]) -> Result<()> {
+        let query = format!(
+            "INSERT INTO `{}` (`{}`, `{}`) VALUES (:path, :value) ON DUPLICATE KEY UPDATE `{}` = :new_value",
+            self.table, self.key_field, self.value_field, self.value_field
+        );
+        let mut conn = self.connection_pool.get_conn().await.map_err(|err| {
+            Error::new(ErrorKind::Unexpected, "connection failed")
+                .with_context("service", Scheme::Mysql)
+                .set_source(err)
+        })?;
+        let statement = conn.prep(query).await.map_err(|err| {
+            Error::new(ErrorKind::Unexpected, "prepare statement failed")
+                .with_context("service", Scheme::Mysql)
+                .set_source(err)
+        })?;
+
+        conn.exec_drop(
+            statement,
+            params! {
+                "path" => path,
+                "value" => value,
+                "new_value" => value,
```

Comment:
Can we avoid encoding value twice in SQL?

Reply:
Seems we can use UPDATE xx = values(xx) instead of it.